File: epi_judge_python/int_as_list_add.py
from typing import Optional

from list_node import ListNode
from test_framework import generic_test


def add_two_numbers(L1: ListNode, L2: ListNode) -> Optional[ListNode]:
    if not L1 and not L2:
        return None
    if not L1:
        return L2
    if not L2:
        return L1
    it = result = ListNode(0, None)
    carry = 0
    while L1 and L2: 
        r = carry + L1.data + L2.data
        carry = r // 10
        it.next = ListNode(r % 10, None)
        it = it.next
        L1 = L1.next
        L2 = L2.next
    while L1:
        r = carry + L1.data
        carry = r // 10
        it.next = ListNode(r % 10, None)
        it = it.next
        L1 = L1.next
    while L2:
        r = carry + L2.data
        carry = r // 10
        it.next = ListNode(r % 10, None)
        it = it.next
        L2 = L2.next
    if carry:
        it.next = ListNode(carry, None)

    return result.next


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('int_as_list_add.py',
                                       'int_as_list_add.tsv', add_two_numbers))

File: epi_judge_cpp/int_as_list_add.cc
#include "list_node.h"
#include "test_framework/generic_test.h"
#include <memory>

shared_ptr<ListNode<int>> AddTwoNumbers(shared_ptr<ListNode<int>> L1,
                                        shared_ptr<ListNode<int>> L2) {
  auto l{L1}, r{L2};
  auto dummy = std::make_shared<ListNode<int>>(), tail{dummy};

  int carry = 0;
  auto append = [&tail, &carry](int sum) {
    carry = sum / 10;
    tail->next = std::make_shared<ListNode<int>>(sum % 10);
    tail = tail->next;
  };

  while (l || r) {
    append(carry + (l ? l->data : 0) + (r ? r->data : 0));
    if (l) {
      l = l->next;
    }
    if (r) {
      r = r->next;
    }
  }
  if (carry) {
    append(carry);
  }
  return dummy->next;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L1", "L2"};
  return GenericTestMain(args, "int_as_list_add.cc", "int_as_list_add.tsv",
                         &AddTwoNumbers, DefaultComparator{}, param_names);
}
